src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs

```diff
@@ -134,18 +150,76 @@ private static unsafe void CopySlow(Array sourceArray, int sourceIndex, Array de
             }
         }
 
-        // Must match the definition in arraynative.cpp
-        private enum AssignArrayEnum
+        private enum ArrayAssignType
         {
-            AssignWrongType,
-            AssignMustCast,
-            AssignBoxValueClassOrPrimitive,
-            AssignUnboxValueClass,
-            AssignPrimitiveWiden,
+            SimpleCopy,
+            WrongType,
+            MustCast,
+            BoxValueClassOrPrimitive,
+            UnboxValueClass,
+            PrimitiveWiden,
         }
 
-        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "Array_CanAssignArrayType")]
-        private static unsafe partial AssignArrayEnum CanAssignArrayType(void* srcTH, void* dstTH);
+        private static unsafe ArrayAssignType CanAssignArrayType(Array sourceArray, Array destinationArray)
```

Comment:
This is much more expensive than it used to be in number of cases. For example, `Array.Copy(src, dst, 1)` where `src` is `new string[1]` and `dst` is `new object[1]` is ~1.2x slower.

I do not see anything particularly wrong with the current QCall. It does not use any problematic patterns. Unless pushing it to C# improves something, I would leave it as is.

Reply:
It's potentially sharable with NativeAOT, though NativeAOT currently doesn't use similar structure.

The entire slow path of `Array.CopyImpl` may get refactored in favor of size in another  way, for example sharing a loop and switch on assignment type in each iteration.

I'll explore the performance first.

Reply:
Well there are other fixes that depend on this PR. It should be better to revert the QCall and merge this first.

Reply:
> The entire slow path of Array.CopyImpl may get refactored in favor of size in another way, for example sharing a loop and switch on assignment type in each iteration.

It would be non-trivial to justify that it is a good trade-off. There are likely apps out there that use the slow path quite a bit.

Reply:
The original purpose of touching the QCall is to combine the copy type detections, since they were originally from one source and have some same logic. It also improves the performance for slow path.

For the string->object case, seperating them doesn't help the performance of this path.

My results:

```csharp
        private string[] strArray = ["a"];
        private object[] objArray = ["a"];

        [Benchmark]
        public void SimpleCopy() => Array.Copy(strArray, objArray, 1);

        [Benchmark]
        public void MustCast() => Array.Copy(objArray, strArray, 1);
```

| Method     | Job        | Toolchain                  | Mean      | Error     | StdDev    | Ratio |
|----------- |----------- |--------------------------- |----------:|----------:|----------:|------:|
| SimpleCopy | Job-MSNWOY | \NoElementType\corerun.exe |  9.748 ns | 0.0814 ns | 0.0721 ns |  0.83 |
| SimpleCopy | Job-NFKZYI | \PR\corerun.exe            | 13.271 ns | 0.0655 ns | 0.0612 ns |  1.14 |
| SimpleCopy | Job-GHUDBI | \main\corerun.exe          | 11.691 ns | 0.0443 ns | 0.0414 ns |  1.00 |
|            |            |                            |           |           |           |       |
| MustCast   | Job-MSNWOY | \NoElementType\corerun.exe | 11.867 ns | 0.0279 ns | 0.0261 ns |  0.50 |
| MustCast   | Job-NFKZYI | \PR\corerun.exe            | 14.467 ns | 0.1299 ns | 0.1215 ns |  0.61 |
| MustCast   | Job-GHUDBI | \main\corerun.exe          | 23.513 ns | 0.0889 ns | 0.0788 ns |  1.00 |

The performance difference is caused by the 2 FCalls of `GetVerifierCorElementType`. It's indeed solvable by touching method table bits instead, but I'm hesitate to mix that with mostly mechanical conversion PR.

Reply:
I would like to avoid introducing perf regressions. Could you please revert the problematic parts from this PR and leave them for follow up?

Reply:
It would be more about a half. I agree this PR could be splitted to unblock individual cases.